benchmarks/profiler_heap_sampling.rb

```diff
@@ -0,0 +1,183 @@
+# Used to quickly run benchmark under RSpec as part of the usual test suite, to validate it didn't bitrot
+VALIDATE_BENCHMARK_MODE = ENV['VALIDATE_BENCHMARK'] == 'true'
+
+return unless __FILE__ == $PROGRAM_NAME || VALIDATE_BENCHMARK_MODE
+
+require 'benchmark/ips'
+require 'ddtrace'
+require 'pry'
+require_relative 'dogstatsd_reporter'
+
+METRIC_VALUES = { 'cpu-time' => 0, 'cpu-samples' => 0, 'wall-time' => 0, 'alloc-samples' => 1, 'timeline' => 0 }.freeze
+OBJECT_CLASS = 'object'.freeze
+
+def sample_object(recorder, depth = 0)
+  if depth <= 0
+    Datadog::Profiling::StackRecorder::Testing._native_track_object(
+      recorder,
+      Object.new,
+      1,
+      OBJECT_CLASS,
+    )
+    Datadog::Profiling::Collectors::Stack::Testing._native_sample(
+      Thread.current,
+      recorder,
+      METRIC_VALUES,
+      [],
+      [],
+      400,
+      false
+    )
+  else
+    sample_object(recorder, depth - 1)
+  end
+end
+
```

Comment:
I'm thinking this benchmark may be a bit... inaccurate in its current form. That's because "no heap" is basically sampling the same thing over and over again, which due to deduping means it's basically bumping a counter somewhere inside libdatadog and thus every iteration is truly the same.

BUT the "heap" variant is actually tracking more and more and more objects. Which means that every iteration is probably going to be slightly slower than the last one, and thus it's kinda hard to make an apples-to-apples comparison. In fact, making the code go faster may make it go slower, because it will allocate even more and thus have more data to wrangle with -- something like that?

I took a peek at the benchmark/ips docs to see if there was some way to fix the number of times the benchmark runs, instead of the time. Because if we said "go for X iterations", at least we could have some comparison of how much time X samples take and it would go lower with each optimization. I didn't find an easy way of doing it.

Hmmm... Thoughts?

(Edit: I guess we could add a `_native_reset_heap_recorder` or something like that that did the simplest thing it could to "untrack" the object and the stack?)

Reply:
Hmm yeah great point :/

We also face a similar issue with the serialization side of things where subsequent iterations are naturally less heavy because the previous one cleared a series of objects.

May need to do some of this stuff outside ips as it feels like we'd want to either support fixing iterations to a set number of enforce the execution of a `before_iteration` that doesn't actively count towards timings.